src/main/java/com/google/cloud/teleport/templates/PubsubToText.java

```diff
@@ -166,11 +173,40 @@ public static PipelineResult run(Options options) {
                         options.getOutputShardTemplate(),
                         options.getOutputFilenameSuffix()))
                 .withTempDirectory(NestedValueProvider.of(
-                    options.getOutputDirectory(),
+                    maybeUseUserTempLocation(
+                        options.getUserTempLocation(),
+                        options.getOutputDirectory()),
                     (SerializableFunction<String, ResourceId>) input ->
                         FileBasedSink.convertToFileResourceIfPossible(input))));
 
     // Execute the pipeline and return the result.
     return pipeline.run();
   }
+
+  /**
+   * Utility method for using optional parameter userTempLocation as TempDirectory.
+   * This is useful when output bucket is locked and temporary data cannot be deleted.
+   *
+   * @param userTempLocation user provided temp location
+   * @param outputLocation default temp location, which should OutputDirectory for backward compatibility
+   * @return userTempLocation if available, otherwise outputLocation is returned.
+   */
+  private static ValueProvider<String> maybeUseUserTempLocation(
+      ValueProvider<String> userTempLocation,
+      ValueProvider<String> outputLocation) {
+    return DualInputNestedValueProvider.of(
+        userTempLocation,
+        outputLocation,
+        new SerializableFunction<TranslatorInput<String, String>, String>() {
+          @Override
+          public String apply(TranslatorInput<String, String> input) {
+            String userProvidedTempLocation = input.getX();
+            String outputLocation = input.getY();
+            if (userProvidedTempLocation == null) {
+              return outputLocation;
+            }
+            return userProvidedTempLocation;
+          }
+        });
```

Comment:
Can this be changed into a simple lambda where you return something like:
(input.getX() != null) ? input.getX() : input.getY();

Reply:
👍